libsawtooth/src/artifact/artifact_creator.rs

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2022 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use crate::error::InternalError;
+
+/// An artifact creator.
+///
+/// Used to create new instances of `Artifact`.
+pub trait ArtifactCreator: Send {
+    /// The context that contains extraneous information required to create the
+    /// `Artifact`.
+    type Context;
+    /// The type of input for the specific `Artifact`
+    type Input;
+    /// The `Artifact` type
+    type Artifact;
+
+    /// Creates a new `Artifact`
+    ///
+    /// Returns a an `Artifact` created from the context and input
+    fn create(
+        &self,
+        context: &mut Self::Context,
+        input: Self::Input,
+    ) -> Result<Self::Artifact, InternalError>;
+}
+
+/// An artifact creator factory.
+///
+/// Used to create new instances of `ArtifactCreator`.
+pub trait ArtifactCreatorFactory {
```

Comment:
Not currently, with the design for the publisher found here https://github.com/hyperledger/sawtooth-lib/pull/151/commits/1efda2418ee134b8ff94b7df4a272f31fc935a43#diff-63fe1738a5d4303b0508f01a7b9b42564c8195208318b8df0b0cdc173bca5d98R113 because currently the ArtifactCreator is Send. However, both could be sent or switch the send to be on the factory, that that would require enforcing Clone on the factory trait.